Move the super() call to the end of FunctionClass.__init__

asmeurer opened the pull request:
The `super()` call calls `BasicMeta.__init__`, which adds the class to a set of
classes (`sympy.core.core.all_classes`). This computes the hash of the class.
However, the hash of a Function depends on its `nargs` (`class_key()`), meaning if
it is set dynamically, the hash will change and the class won't be found in
the `all_classes` set any more, breaking `sympify(obj)`.

This happens for things like `Function('f', eval=lambda self, t: ...)` where
`nargs` is determined from the `eval` lambda.


<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234". See
https://github.com/blog/1506-closing-issues-via-pull-requests .-->

Fixes #15170.

#### Brief description of what is fixed or changed


#### Other comments


#### Release Notes

<!-- Write the release notes for this release below. See
https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more information
on how to write release notes. The bot will check your release notes
automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
* core
  * Fix an issue with undefined functions defined with a custom `eval` (like `Function('f', eval=...)`).
<!-- END RELEASE NOTES -->

asmeurer committed:
Move the super() call to the end of FunctionClass.__init__

The super() call calls BasicMeta.__init__, which adds the class to a set of
classes (sympy.core.core.all_classes). This computes the hash of the class.
However, the hash of a Function depends on its nargs (class_key()), meaning if
it is set dynamically, the hash will change and the class won't be found in
the all_classes set any more, breaking sympify(obj).

This happens for things like Function('f', eval=lambda self, t: ...) where
nargs is determined from the eval lambda.

Fixes #15170.

asmeurer committed:
Fix a test so it works in Python 2

jksuom commented on sympy/core/function.py:
```diff
@@ -162,6 +161,8 @@ def __init__(cls, *args, **kwargs):
             nargs = (as_int(nargs),)
         cls._nargs = nargs
 
+        super(FunctionClass, cls).__init__(args, kwargs)
```
It looks like ManagedProperties and BasicMeta will read this as two args (and empty keyword dict) but it does not matter as they do not use them.

asmeurer replied:
Oh I didn't notice that. Obviously it should be `*args, **kwargs`.

asmeurer committed:
Fix arguments to super().__init__

The aren't actually used, but obviously it should be __init__(*args,
**kwargs), not __init__(args, kwargs).

jksuom commented:
Restarted a non-started job.
